libraries/AP_Baro/AP_Baro_UAVCAN.cpp

```diff
@@ -3,100 +3,175 @@
+    const int pressure_listener_res = pressure_listener->start(PressureCb(ap_uavcan, &handle_pressure));                     //Msg Handler
+    if (pressure_listener_res < 0) {
+        AP_HAL::panic("UAVCAN Baro subscriber start problem\n\r");
         return;
     }
 
-    ap_uavcan->remove_baro_listener(this);
+    uavcan::Subscriber<uavcan::equipment::air_data::StaticTemperature, TemperatureCb> *temperature_listener;
+    temperature_listener = new uavcan::Subscriber<uavcan::equipment::air_data::StaticTemperature, TemperatureCb>(*node);
+    const int temperature_listener_res = temperature_listener->start(TemperatureCb(ap_uavcan, &handle_temperature));                          //Msg Handler
+    if (temperature_listener_res < 0) {
+        AP_HAL::panic("UAVCAN Baro subscriber start problem\n\r");
+        return;
+    }
+}
 
-    debug_baro_uavcan(2, _manager, "AP_Baro_UAVCAN destructed\n\r");
+void AP_Baro_UAVCAN::handle_pressure(AP_UAVCAN* ap_uavcan, uint8_t node_id, const PressureCb &cb)
+{
+    AP_Baro_UAVCAN* driver = get_uavcan_backend(ap_uavcan, node_id, true);
+    if (driver == nullptr) {
+        return;
+    }
+    WITH_SEMAPHORE(driver->_sem_baro);
+    driver->_pressure = cb.msg->static_pressure;
+    driver->new_pressure = true;
 }
 
-AP_Baro_Backend *AP_Baro_UAVCAN::probe(AP_Baro &baro)
+void AP_Baro_UAVCAN::handle_temperature(AP_UAVCAN* ap_uavcan, uint8_t node_id, const TemperatureCb &cb)
 {
-    uint8_t can_num_drivers = AP::can().get_num_drivers();
+    AP_Baro_UAVCAN* driver = get_uavcan_backend(ap_uavcan, node_id, false);
+    if (driver == nullptr) {
+        return;
+    }
+    WITH_SEMAPHORE(driver->_sem_baro);
+    driver->_temperature = cb.msg->static_temperature;
+}
 
-    AP_Baro_UAVCAN *sensor;
-    for (uint8_t i = 0; i < can_num_drivers; i++) {
-        AP_UAVCAN *ap_uavcan = AP_UAVCAN::get_uavcan(i);
-        if (ap_uavcan == nullptr) {
-            continue;
-        }
 
-        uint8_t freebaro = ap_uavcan->find_smallest_free_baro_node();
-        if (freebaro == UINT8_MAX) {
-            continue;
+
+AP_Baro_UAVCAN* AP_Baro_UAVCAN::get_uavcan_backend(AP_UAVCAN* ap_uavcan, uint8_t node_id, bool create_new)
+{
+    AP_Baro_UAVCAN* driver = nullptr;
+    if (ap_uavcan == nullptr) {
+        return nullptr;
+    }
+    for (uint8_t i = 0; i < BARO_MAX_DRIVERS; i++) {
+        if (_detected_modules[i].driver != nullptr &&
+            _detected_modules[i].ap_uavcan == ap_uavcan && 
+            _detected_modules[i].node_id == node_id) {
+            return _detected_modules[i].driver;
         }
-        sensor = new AP_Baro_UAVCAN(baro);
-        if (sensor->register_uavcan_baro(i, freebaro)) {
-            debug_baro_uavcan(2, i, "AP_Baro_UAVCAN probed, drv: %d, node: %d\n\r", i, freebaro);
-            return sensor;
-        } else {
-            delete sensor;
+    }
+    
+    if (driver == nullptr && create_new) {
```

Comment:
How can it be that `driver != nullptr` ?

Reply:
Yes, this is unneeded, just an artefact from changes in previous attempts.